sdk/core/Azure.Core/src/ResourceModifiedException.cs

```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure
+{
+    public class ResourceModifiedException : Exception
```

Comment:
Why doesn't this derive from `RequestFailedException`?  I get that we're going to throw this from `Response<T>.Value` which people may feel is too far away from when the request was made, but wouldn't all the response info be helpful for users in figuring out where things went wrong?  This just says a resource was modified, but why wouldn't it tell me which?  And since it's not tied to the response and wasn't raised directly by the operation, does this mean that my request succeeded but then something locally modified my `Response<T>.Value`?

We're being clever with how we throw this so we should be clear about what's being thrown.  It's a specialized request failure so I think it should derive from `RequestFailedException`.  We'd also get the helpful `RequestFailedException.ToString` for free.

If we really don't want to derive from `RequestFailedException`, then we need to introduce a base exception that both `RequestFailedException` and `ResourceModifiedException` derive from.  I want there to be a single thing customers have to catch when they're getting started.  They can add specialized catch clauses once they build expertise and explore further into our features.

Reply:
In this case, the request didn't fail, so it's not a RequestFailedException.  Let me think about what a base class for both would be - open to suggestions here.

Reply:
ResourceModifiedException  is used to indicate 304 (resource not modified)? No confusing about the name.

Reply:
@annelo-msft - I see it as some kind of failure.  I asked you to do a thing and you didn't because of my other conditions, but you still didn't do the thing I wanted.  And you're throwing an exception which I'd say is indicative a failure occurred.  If you don't want to derive from `RequestFailedException`, then we should move it all into `RequestException` or `AzureException` and derive both from that.

@xiangyan99 - I disagree.  I think it's really confusing for .NET developers because they're not as familiar with HTTP _(maybe because ASP.NET hides HTTP from folks)_.  You're hiding the HTTP complexity behind a friendly `ifUnchanged` boolean, but then switching back to the HTTP concept with the exception.

Reply:
@xiangyan99, you're right about naming.  I'll fix that.

Reply:
@tg-msft In a 304, it isn't a failure.  The caller asked the service to get something if it changed.  If it didn't change and I didn't give you an updated value, I did precisely what you asked.  3xx codes mean the caller has to do more work to get the right version (not a failure), and 304 is an odd man out in this group indicating a cache doesn't need to be updated -- the extra work you have to do is to just use the value you already have, rather than re-requesting more information from the service.

I'll move forward with a RequestException base class - I think that's a good proposal.

Reply:
A better name for this exception is "ResponseBodyNotFoundException."  ResourceUnmodifiedException could feasibly inherit from this, but I know we don't want to get into a deep exception hierarchy at this point.

Reply:
Tracking in issue https://github.com/Azure/azure-sdk-for-net/issues/7897